src/transformers/models/superglue/modeling_superglue.py

```diff
@@ -0,0 +1,806 @@
+        indices0, indices1 = max0.indices, max1.indices
+        mutual0 = arange_like(indices0, 1)[None] == indices1.gather(1, indices0)
+        mutual1 = arange_like(indices1, 1)[None] == indices0.gather(1, indices1)
+        zero = scores.new_tensor(0)
+        matching_scores_0 = torch.where(mutual0, max0.values.exp(), zero)
+        matching_scores_1 = torch.where(mutual1, matching_scores_0.gather(1, indices1), zero)
+        valid0 = mutual0 & (matching_scores_0 > self.config.matching_threshold)
+        valid1 = mutual1 & valid0.gather(1, indices1)
+        matches_0 = torch.where(valid0, indices0, indices0.new_tensor(-1))
+        matches_1 = torch.where(valid1, indices1, indices1.new_tensor(-1))
+
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + concat_hidden_states_tuples_pair(
+                encoded_keypoints0[1], encoded_keypoints1[1]
+            )
+            all_hidden_states = all_hidden_states + gnn_outputs[2]
+            all_hidden_states = all_hidden_states + concat_hidden_states_tuples_pair(
+                (projected_descriptors_0,), (projected_descriptors_1,)
+            )
+
+        if output_attentions:
+            all_attentions = all_attentions + gnn_outputs[3]
+
+        return (
+            matches_0,
+            matches_1,
+            matching_scores_0,
+            matching_scores_1,
+            all_hidden_states,
+            all_attentions,
+        )
+
+    @add_start_docstrings_to_model_forward(SUPERGLUE_INPUTS_DOCSTRING)
+    def forward(
+        self,
+        pixel_values: torch.FloatTensor = None,
+        labels: Optional[torch.LongTensor] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+    ) -> Union[Tuple, KeypointMatchingOutput]:
+        loss = None
+        if labels is not None:
+            raise ValueError("SuperGlue is not trainable, no labels should be provided.")
+
+        output_attentions = output_attentions if output_attentions is not None else self.config.output_attentions
+        output_hidden_states = (
+            output_hidden_states if output_hidden_states is not None else self.config.output_hidden_states
+        )
+        return_dict = return_dict if return_dict is not None else self.config.use_return_dict
+
+        if len(pixel_values.size()) != 5 or pixel_values.size(1) != 2:
+            raise ValueError("Input must be a 5D tensor of shape (batch_size, 2, num_channels, height, width)")
+
+        batch_size, _, channels, height, width = pixel_values.shape
+
+        list_matches_0 = []
+        list_matches_1 = []
+        list_matching_scores_0 = []
+        list_matching_scores_1 = []
+        list_keypoints_0 = []
+        list_keypoints_1 = []
+        list_hidden_states = []
+        list_attentions = []
+
+        for i in range(pixel_values.size(0)):
+            image_pair = pixel_values[i]
```

Comment:
We can iterate directly on the tensor 

```suggestion
        for image_pair in pixel_values:
```